### ISSUE TITLE
[bug] `TensorIterator` segfaults when inputs broadcast over output, and output is at least 2 dims larger than inputs.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#79218]
### 🐛 Describe the bug

This one was originally observed in https://github.com/pytorch/pytorch/issues/78867.
To reproduce, checkout https://github.com/pytorch/pytorch/pull/79201 and run

```python
import itertools
import torch

res = [torch.rand((5,) * d, dtype=torch.double, device='cpu') for d in range(4)]

for input in itertools.product(res, repeat=3):
    print(f"shape : {[i.shape for i in input]}")
    torch.foo(*input)

```

This outputs:
```
...
CPU kernel                                                                                                             
shape : [torch.Size([5, 5, 5]), torch.Size([]), torch.Size([])]                                                                                                                                                                                                                                                              CPU kernel                                                                                                                                                                                                                                                                                                                   /home/nik/git/Quansight/pytorch/test.py:8: UserWarning: An output with one or more elements was resized since it had shape [5, 5, 5], which does not match the required output shape []. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  /home/nik/git/Quansight/pytorch/aten/src/ATen/native/Resize.cpp:17.)      
  torch.foo(*input)                                                                                                    
shape : [torch.Size([5, 5, 5]), torch.Size([]), torch.Size([5])]                                                       
CPU kernel                                                                                                             
Segmentation fault (core dumped)
```

This also segfaults on CUDA, which might suggest that the issue is in the TI's internals.
NOTE, however, that the script works just fine if restricted to tensors with dim `< 3`.

A simpler repro with an `out` specified:
```python
In [1]: import torch

In [2]: x = torch.rand(2)

In [3]: y = torch.rand(())

In [4]: z = torch.rand(2, 2, 2)

In [5]: torch.add(x, y, out=z)
Segmentation fault (core dumped)
```
This indicates that, potentially, all out functions are unsafe.

Segfaults happen when `out` is at least 2 dimensions larger than any of the inputs in point-wise-like operations.

### Versions

master

CC @ngimel 

cc @ezyang @gchanan @zou3519

**Comment by zou3519** [comment]
What is "foo"? Any TensorIterator opration?

**Comment by redmagnos** [comment]
I'm gonna try to tackle this if no one else is already looking into it.

**Comment by nikitaved** [comment]
Hey, @redmagnos , sorry for the delay. This warning is about output being broadcasted over input, so you have several elements in the input that correspond to the same element in the output. Should not be a problem in the context of `foo`.

**Comment by redmagnos** [comment]
@nikitaved Oh I know it's not a problem for this case. What I meant was that the warning specifically said that the functionality of resizing the output to match the input like that will no longer be supported in the future, and depending on how that's implemented, it might solve the issue all on its own (since the issue seems to be exactly related to the relative sizes of the tensors). But again, it might also not, which is why I ask what the actual new behaviour in the future will be.

**Comment by redmagnos** [comment]
Right, I understand this; that shape _also_ causes the segfault for me, any shapes where the output is 2+ larger than the inputs and the inputs are not the same size segfault, and the one you suggest is just a special case of this general rule.

However, with the current version of pyTorch I have installed the warning _is_ triggered even in those cases. If what you mean here is that for cases where the output is larger than the inputs the warning will merely be silenced and the behaviour will not change, then I will resume looking into the bug. That is the question I am asking, is what future behaviour changes there will be for this and similar cases, if any.

**Comment by redmagnos** [comment]
Yeah `loop_2d` was my first guess at where to look at and I'll probably get to it todayish.

**Comment by redmagnos** [comment]
Alright so I uh kinda forgot to keep looking at this and took a long break but I'm back.

@nikitaved what is the expected behaviour of `z = torch.rand(2, 2, 2)`? Let's take a random example of:

```
>>> x
tensor([0.7966, 0.7535])
>>> y
tensor(0.7590)
>>> z
tensor([[[0.0570, 0.0964],
         [0.9760, 0.1225]],

        [[0.9648, 0.9186],
         [0.3263, 0.5028]]])
```

Should `z` be overwritten and contain the regular result of `torch.add(x, y)`?
```
>>> z
tensor([1.5557, 1.5125])
```

Or should all rows of it be overwritten instead?
```
>>> z
tensor([[[1.5557, 1.5125],
         [1.5557, 1.5125]],

        [[1.5557, 1.5125],
         [1.5557, 1.5125]]])
```
Or should only the first row be overwritten?
```
>>> z
tensor([[[1.5557, 1.5125],
         [0.9760, 0.1225]],

        [[0.9648, 0.9186],
         [0.3263, 0.5028]]])
```
